File: src/com/medium/LRUCache_146.java
/**
 * 
 */
package com.medium;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class LRUCache_146 {
    // 2019年8月6日
    // Use Linked HashMap
    // Time Complexity: O(1)
    // Space Complexity:O(n)
    // Runtime: 61 ms, faster than 68.53%
    // Memory Usage: 57.3 MB, less than 67.17%
    private LinkedHashMap<Integer, Integer> mMap;
    private int capacity;

    public LRUCache_146(int capacity) {
        this.capacity = capacity;
        mMap = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
            protected boolean removeEldestEntry(Map.Entry eldest) {
                return size() > capacity;
            }
        };
    }

    public int get(int key) {
        return mMap.getOrDefault(key, -1);
    }

    public void put(int key, int value) {
        mMap.put(key, value);
    }
}

// 2019年8月6日
// Use HashMap and double linkedlist
// Time Complexity: O(1)
// Space Complexity:O(n)
// Runtime: 62 ms, faster than 58.49%
// Memory Usage: 57.9 MB, less than 56.99%
class LRUCache {
    class Node {
        int key;
        int value;
        Node pre;
        Node next;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private Node head;
    private Node tail;
    private int capacity;
    private HashMap<Integer, Node> mMap;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        mMap = new HashMap<>();
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head.next = tail;
        tail.pre = head;
        head.pre = null;
        tail.next = null;
    }

    public void addNode2Head(Node mNode) {
        Node mNodeHeadNext = head.next;
        head.next = mNode;
        mNodeHeadNext.pre = mNode;
        mNode.next = mNodeHeadNext;
        mNode.pre = head;
    }

    public void deleteNode(Node mNode) {
        mNode.pre.next = mNode.next;
        mNode.next.pre = mNode.pre;
    }

    public int get(int key) {
        if (mMap.containsKey(key)) {
            Node mNode = mMap.get(key);
            int mResult = mNode.value;
            deleteNode(mNode);
            addNode2Head(mNode);

            return mResult;
        }
        return -1;
    }

    public void put(int key, int value) {
        if (mMap.containsKey(key)) {
            Node mNode = mMap.get(key);
            deleteNode(mNode);
            mNode.value = value;
            addNode2Head(mNode);
        } else {
            if (mMap.size() == capacity) {
                mMap.remove(tail.pre.key);
                deleteNode(tail.pre);
            }
            Node mNode = new Node(key, value);
            addNode2Head(mNode);
            mMap.put(key, mNode);

        }
    }

}

File: Golang_src/medium/LRUCache_146.go
package medium

// Use hash table + double linkedlist:
// Time Complexity: O(1)
// Space Complexity:O(capacity)
// Runtime: 124 ms, faster than 72.53%
// Memory Usage: 17.2 MB, less than 41.57%
type LRUCache struct {
	cache      map[int]*DoubleLinkedNode
	size       int
	capacity   int
	head, tail *DoubleLinkedNode
}

type DoubleLinkedNode struct {
	key, val  int
	pre, next *DoubleLinkedNode
}

func createDoubleLinkedNode(key, val int) *DoubleLinkedNode {
	return &DoubleLinkedNode{
		key: key,
		val: val,
	}
}

func CreateLRU(capacity int) LRUCache {
	l := LRUCache{
		cache:    map[int]*DoubleLinkedNode{},
		head:     createDoubleLinkedNode(0, 0),
		tail:     createDoubleLinkedNode(0, 0),
		capacity: capacity,
	}

	l.head.next = l.tail
	l.tail.next = l.head

	return l
}

func (this *LRUCache) Get(key int) int {
	if _, exist := this.cache[key]; !exist {
		return -1
	}
	node := this.cache[key]
	this.moveToHead(node)

	return node.val
}

func (this *LRUCache) Put(key int, value int) {
	if _, exist := this.cache[key]; !exist {
		node := createDoubleLinkedNode(key, value)
		this.cache[key] = node
		this.addToHead(node)
		this.size++
		if this.size > this.capacity {
			remove := this.removeTail()
			delete(this.cache, remove.key)
			this.size--
		}
	} else {
		node := this.cache[key]
		node.val = value
		this.moveToHead(node)
	}
}

func (this *LRUCache) addToHead(node *DoubleLinkedNode) {
	node.pre = this.head
	node.next = this.head.next
	this.head.next.pre = node
	this.head.next = node
}

func (this *LRUCache) removeNode(node *DoubleLinkedNode) {
	node.pre.next = node.next
	node.next.pre = node.pre
}

func (this *LRUCache) moveToHead(node *DoubleLinkedNode) {
	this.removeNode(node)
	this.addToHead(node)
}

func (this *LRUCache) removeTail() *DoubleLinkedNode {
	node := this.tail.pre
	this.removeNode(node)

	return node
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
